### ISSUE TITLE
[Feature]: separate termux for root

### ISSUE THREAD

**Issue by beroal** [termux/termux-app#2536]
### Feature description

If I understand it correctly, any program executed in the Termux shell can write to `$PREFIX/bin`. If any Termux program (say, `$PREFIX/bin/$E`) is executed as root, then **any** Termux program has root privileges by infecting `$PREFIX/bin/$E`. In this case, a user should only execute programs from a very restricted set for security reasons. This reduces the usefulness of Termux.

If Termux programs are executed as root only in the separate Android package, then Termux programs executed in the main package can't obtain root privileges since Android packages are isolated.

Therefore, having a separate Termux package for root will increase the usefulness of Termux without compromising security.

### Additional information

Obviously, Linux has this feature. In Linux, an ordinary user can't write to `/bin`.

**Comment by beroal** [comment]
> How would this work do you mean? We have a termux clone and keep two copies of every program, one for root usage and one for normal user?

I guess this is required because the ID of the Termux application is hardcoded in every Termux package. You know better than me ☺️.

Even if the root Termux application requires separate compilation of Termux programs, only programs that have a function only usable as root need to be compiled, basically, `root-repo`.

> Root copies would still be able to "infect" other "root programs" in that case

I don't see what this gives to an attacker. A user would give the same (maximal) level of privileges to all programs in the root Termux application anyway.

The benefit lies in isolating Termux programs that a user wants to execute as root from other Termux programs.

> You might also be interested in the work adapting termux for [android 10](https://github.com/termux/termux-packages/wiki/Termux-and-Android-10), that requires packaging all programs into APK(s), and gives read only programs

Yes, I'm reading about this now. This may be a solution. But I heard that all these APKs can write each other's private data. If the root Termux application contains a shell script, that script may be infected. Infection trough configuration files is possible too. For example, if a configuration file contains the path to an executable.

> That is a wise, or better yet: don't root your phone at all.

Unfortunately, some Android functions are only accessible to root. You have a whole repository of such programs ☺️!

I believe that Android can provide a higher level of security than Linux due to its application isolation and Storage Access Framework. At least, it shouldn't be worse than Linux in the root aspect since it also has the Unix concept of user accounts.

**Comment by Grimler91** [comment]
The packages we have in termux-root-packags are all dynamically linked against other libraries, so to fully isolate them they would also need to be statically linked instead, or have all the dependencies be part of the root app as well (basically meaning fork termux-app and rebuilding relevant packages with new $PREFIX)

**Comment by magbulogtong81** [comment]
@beroal ofcourse users cannot change `/bin` since its owned by root UID and UUID. in termux (or all apps basically), whatever uniqut user you got (e.g. u0_a625), it's the same for every files and folder. but, AFAIK termux packages will break only if you use installers (e.g. apt, python pip, etc) with root. I'm speaking from experience. i have some rooted android device and it's working just fine with termux. just make sure you're not on root when you install something

**Comment by beroal** [comment]
> but, AFAIK termux packages will break only if you use installers (e.g. apt, python pip, etc) with root. I'm speaking from experience. i have some rooted android device and it's working just fine with termux.

Do you propose to make root the owner of `$PREFIX`? This is quite interesting.

**Comment by agnostic-apollo** [comment]
> Do you propose to make root the owner of `$PREFIX`? This is quite interesting.

Most definitely not. The termux app will not be able to access `$PREFIX` and will not work at all. The ownership, permissions and selinux context must never be messed around with, that is exactly why termux adds a condition in `apt` to not allow users to run as root, since users mess their environment and then complain to termux. Same reason any files under `~/` home must not be messed around with, wrapper scripts for root like [`sudo`](https://github.com/agnostic-apollo/sudo) use a separate home `~/.suroot` for this.

What you are asking as mentioned above is possible with a separate forked app with a different package name considering superuser managers like Magisk grant universal access to the app. There is no password prompts either like in linux distros, but there is an ask every time prompt, so no way to separate inside a single app. There is already a Termux:Sudo issue #619 open. Even for that a separate app that does not share a `sharedUserId` with termux would be required. 

However, for rooted users, future support may be added for multi user support, so you can install a trusted termux in secondary user, so changing package name would not be required. Check https://github.com/termux/termux-app/issues/2168#issuecomment-877621546.

**Comment by beroal** [comment]
> Most definitely not. The termux app will not be able to access `$PREFIX` and will not work at all. The ownership, permissions and selinux context must never be messed around with

Of course, you can grant read-only access to the Termux application to the files owned by root. I don't know how SELinux plays into this since I'm not familiar with it.

> There is no password prompts either like in linux distros, but there is an ask every time prompt, so no way to separate inside a single app.

 I don't see how you came to this conclusion. Passwords are only needed if there is more than one user which isn't the case here. How Magisk grants permissions is how it should be done in Linux.

**Comment by agnostic-apollo** [comment]
>you can grant read-only access to the Termux application to the files owned by root.

Any malicious program in the non-root area can just get write access too if needed through magisk.

>I don't know how SELinux plays into this since I'm not familiar with it.

Messing around with `root` changes the selinux context of files and termux app can't access them, even if it has ownership and permission.

https://github.com/termux/termux-app/issues/2080#issuecomment-846575028

Both linux distros and magisk grant access per user. In linux distros, running `sudo` asks for a password whenever you run it, with a timeout. Magisk either grants access forever, once, or with timeout, but never asks for any password. You can't filter requesting access from inside the same single termux user, need separate users.

>Passwords are only needed if there is more than one user

Passwords are by default asked for non-root users in sudo group, and linux distros do have at least one of those in which you log in. You can optionally disable it if you want with `NOPASSWD`.

https://askubuntu.com/questions/147241/execute-sudo-without-password

**Comment by beroal** [comment]
I'll wait for this feature that allows to run Termux under a secondary user. I don't agree with your arguments, but I won't argue. Thank you for your work. Where can I track the progress of this feature?

**Comment by agnostic-apollo** [comment]
It's all chill. Anyways, welcome. It is planned to be added but lot of other higher priority work is going on that needs to completed first before I can get to it. It also requires lot of other related work to be done since other variants needs to be supported too, so designing it will take time.

**Comment by twaik** [comment]
Seems like this issue can be closed, right?
